doc/man/gnunet-download.1
.TH GNUNET-DOWNLOAD "1" "16 Jun 2007" "GNUnet"
.SH NAME
gnunet\-download \- a command line interface for downloading files from GNUnet
.SH SYNOPSIS
.B gnunet\-download
[\fIOPTIONS\fR] \-\- GNUNET_URI
.SH DESCRIPTION
.PP
Download files from GNUnet.
.TP
\fB\-a \fILEVEL\fR, \fB\-\-anonymity=LEVEL\fR
set desired level of receiver anonymity.  Default is 1.
.TP
\fB\-c \fIFILENAME\fR, \fB\-\-config=FILENAME\fR
use config file (defaults: ~/.gnunet/gnunet.conf)
.TP
\fB\-d, \fB\-\-directory\fR
download a GNUnet directory that has already been downloaded.  Requires that a filename of an existing file is specified instead of the URI.  The download will only download the top\-level files in the directory unless the `-R' option is also specified.
.TP
\fB\-D, \fB\-\-delete-incomplete\fR
causes gnunet\-download to delete incomplete downloads when aborted with CTRL-C.  Note that complete files that are part of an incomplete recursive download will not be deleted even with this option.  Without this option, terminating gnunet\-download with a signal will cause incomplete downloads to stay on disk.  If gnunet\-download runs to (normal) completion finishing the download, this option has no effect.
.TP
\fB\-h\fR, \fB\-\-help\fR
print help page
.TP
\fB\-H \fIHOSTNAME\fR, \fB\-\-host=\fIHOSTNAME\fR
on which host is gnunetd running (default: localhost).  You can also specify a port using the syntax HOSTNAME:PORT.  The default port is 2087.
.TP
\fB\-L \fILOGLEVEL\fR, \fB\-\-loglevel=LOGLEVEL\fR
Change the loglevel.  Possible values for LOGLEVEL are NOTHING, FATAL,
ERROR, WARNING, INFO, STATUS and DEBUG.  Note that options in the 
configuration file take precedence over this option (the argument 
will be ignored in that case).
.TP
\fB\-o \fIFILENAME\fR, \fB\-\-output=FILENAME\fR
write the file to FILENAME.  Hint: when recursively downloading a directory, append a '/' to the end of the FILENAME to create a directory of that name.  If no FILENAME is specified, gnunet\-download constructs a temporary ID from the URI of the file.  The final filename is constructed based on meta\-data extracted using libextractor (if available).
.TP
\fB\-p \fIDOWNLOADS\fR, \fB\-\-parallelism=DOWNLOADS\fR
set the maximum number of parallel downloads that is allowed.  More parallel downloads can, to some extent, improve the overall time to download content.  However, parallel downloads also take more memory.  The specified number is the number of files that are downloaded in parallel, not the number of blocks that are concurrently requested.  As a result, the number only matters for recursive downloads.  The default value is 32.
.TP
\fB\-R\fR, \fB\-\-recursive\fR
download directories recursively (and in parallel); note that the URI must belong to a GNUnet directory and that the filename given must end with a '/' -- otherwise, only the file corresponding to the URI will be downloaded.
.TP
\fB\-v\fR, \fB\-\-version\fR
print the version number
.TP
\fB\-V\fR, \fB\-\-verbose\fR
print progress information
.SH NOTES
As most GNUnet command\-line tools, gnunet\-download supports passing arguments using environment variables.  This can improve your privacy since otherwise the download URI will likely be visible to other local users.  Setting "GNUNET_ARGS" will cause the respective string to be appended to the actual command-line and to be processed the same way as arguments given directly at the command line.

The GNUNET_URI is typically obtained from gnunet\-search. gnunet\-gtk can also be used instead of gnunet\-download.
If youever have to abort a download, you can at any time continue it by re\-issuing gnunet\-download with the same filename. In that case GNUnet will not download blocks again that are already present. GNUnets file\-encoding will ensure file integrity, even if the existing file was not downloaded from GNUnet in the first place. Temporary information will be stored in FILENAME.X files until the download is completed. These files are used only if a download is resumed later. If you abort a download for good, you should remember to delete these files.

.SH SETTING ANONYMITY LEVEL

The \fB\-a\fR option can be used to specify additional anonymity constraints. If set to 0, GNUnet will try to download the file as fast as possible without any additional slowdown by the anonymity code. Note that you will still have a fair degree of anonymity depending on the current network load and the power of the adversary. The download is still unlikely to be terribly fast since the sender may have requested sender\-anonymity and since in addition to that, GNUnet will still do the anonymous routing.

This option can be used to limit requests further than that. In particular, you can require GNUnet to receive certain amounts of traffic from other peers before sending your queries. This way, you can gain very high levels of anonymity \- at the expense of much more traffic and much higher latency. So set it only if you really believe you need it.

The definition of ANONYMITY\-RECEIVE is the following: If the value v is < 1000, it means that if GNUnet routes n bytes of messages from foreign peers, it may originate n/v bytes of queries in the same time\-period. The time\-period is twice the average delay that GNUnet deferrs forwarded queries. If the value v is >= 1000, it means that if GNUnet routes n bytes of QUERIES from at least (v % 1000) peers, it may originate n/v/1000 bytes of queries in the same time\-period.

The default is 0 and this should be fine for most users. Also notice that if you choose values above 1000, you may end up having no throughput at all, especially if many of your fellow GNUnet\-peers do the same.

.SH FILES
.TP
~/.gnunet/gnunet.conf
GNUnet configuration file
.SH "REPORTING BUGS"
Report bugs by using mantis <https://gnunet.org/bugs/> or by sending electronic mail to <gnunet\-developers@gnu.org>
.SH "SEE ALSO"
\fBgnunet\-gtk\fP(1), \fBgnunet\-insert\fP(1), \fBgnunet\-gtk\fP(1), \fBgnunet\-search\fP(1), \fBgnunet\-download\fP(1), \fBgnunet.conf\fP(5), \fBgnunetd\fP(1)


ChangeLog
Mon Nov 16 09:25:47 CET 2009
	Allow passing of command-line arguments using GNUNET_ARGS
	environment variable (see Mantis #1467).

Mon Apr 27 22:18:19 CEST 2009
	Monitor low-traffic connections over a longer period
	of time before terminating them.

Fri Feb 27 22:05:50 MST 2009
	Releasing 0.8.0c.

Sun Feb  1 19:51:40 MST 2009
	Fixed Mantis #1429 (struct padding could cause sizeof
	to be different on AMD64, using gcc-ism "packed" to
	avoid this problem).

Sun Feb  1 00:37:16 MST 2009
	Fixed problem in DHT client protocol that could cause
	PUT requests from clients to be (frequently) lost.

Sun Nov  2 16:06:47 MST 2008
	Fixed problem with MySQL disconnects in one module
	impacting MySQL correctness in another module (caused
	crashes).

Sat Aug 16 00:18:34 CEST 2008
	Improved keyword extraction
	Fixed expiration time of keywords
	Releasing 0.8.0b.

Sun Jul 20 15:27:11 MDT 2008
	Releasing 0.8.0a.

Thu Jul 10 00:59:31 MDT 2008
	Fixed bug where datastore resizing would not 
	result in bloomfilter update during gnunet-update.

Sat Jun 28 23:43:00 MDT 2008
	Fixed bug preventing namespace advertisements from
	being processed.

Sat Jun 14 00:54:39 MDT 2008
	Releasing 0.8.0.

Fri Jun 13 23:51:00 MDT 2008
	GNUnet now compiles on OpenBSD.

Wed Jun 11 22:17:09 MDT 2008
	Made state of gnunet-auto-share persistent. 
	Added support for multiple directories.  Improved
	FSUI/ECRS APIs to reduce the number of threads needed
	for probing.  Increased number of concurrent probes
	allowed.

Sat Jun  7 17:54:49 MDT 2008
	Simplified namespaces a lot.  Shorter timeouts for
	trying to connect to gnunetd to prevent interactive
	apps from blocking too long when gnunetd is not
	running.

Sun Jun  1 21:11:54 MDT 2008
	Releasing 0.8.0pre1.

Sun Jun  1 13:41:27 MDT 2008
	Split of libgnunetpseudonym from libgnunetnamespace
	(with major API extension and changes).

Tue May  6 04:10:02 MDT 2008
	Added simple sanity check for [MODULES] values.
	Added "-s" and "-q" options to gnunet-peer-info.

Thu Apr 24 00:05:17 MDT 2008
	Simplified FSUI stop/abort/pause/resume API calls
	(no need to pass FSUI context anymore).

Mon Apr 21 21:05:20 MDT 2008
	Adding buffered IO for FSUI's serialize and
	deserialize code.

Sun Apr 20 20:01:20 MDT 2008
	Fixed bug in HTTP transport causing lonely messages
	(responses to GET) not to be delivered.

Mon Mar 24 21:15:36 MDT 2008
	Fixed various problems with downloading locally
	indexed large files (downloading large files from
	other peers should not have been a problem).
	
Sun Mar  2 15:33:41 MST 2008
	Added option to allow user to disable IPv6 support.
	Releasing 0.8.0pre0.

Sat Mar  1 21:14:02 MST 2008
	Fixed some dstore performance issues.
	Made transports alternate between trying IPv4 and IPv6
	if both are available (since one of them maybe
	misconfigured).

Mon Feb 25 23:25:48 MST 2008
	Cleanup of the DNS code: removed GNUNET_IPvXAddress
	and GNUNET_PRIP; centralized all DNS lookup code in 
	util/network/dns.c; improved IPv6 support for 
        network-client code.

Mon Feb 25 00:01:27 MST 2008
	Added asynchronous search and download methods for 
	ECRS library.  FSUI now can do with only one thread
	per search or download (until now, we had two 
	threads per search / download).

Tue Feb 19 20:35:28 MST 2008
	Updated database schemata to support O(1) operations
	even if there are N files under the same keywords.
	This breaks database backwards compatibility (and
	no migration code is provided).

Mon Feb 18 19:47:37 MST 2008
	Added new options -u and -s to gnunet-insert.

Sat Feb 16 21:37:33 MST 2008
	Implemented gnunet-auto-share for automatic
	sharing of directories.

Thu Feb 14 01:02:32 MST 2008
	Made HTTP transport work over IPv6 as well.

Wed Feb 13 22:59:07 MST 2008
	Modified mySQL and SQLite datastores to return diverse
	result sets during (partial) hash-based get iterations
	(should result in more diverse keyword result sets).

Tue Feb 12 23:54:34 MST 2008
	First shot at integrated hostlist HTTP server based
	on libmicrohttpd.

Sat Feb  9 22:06:23 MST 2008
	Minor changes to the ECRS/FSUI APIs for searches
	(no more timeouts and/or max-results).  Clients who
	need these features could easily implement them 
	directly (and they are not really useful to begin
	with).  
	Also, the IPv4 and IPv6 protocols are being integrated
	so that we only listen to one port (running both IPv4
	and IPv6 over the same port).  This break transport
	compatibility, but given that we're breaking protocol
	compatibility for file-sharing already, this may 
	actually be a bit of a good thing (since it will now
	be possible to tell if peers running the new GAP code
	are connected to other peers running the new GAP code).
	
Sun Feb  3 13:17:09 MST 2008
	Dramatic changes to the GAP implementation (breaking
	protocol compatibility).  Essentially, we can save
	a few bytes in each reply.  More importantly, the new
	code allows the searching client to specify a set of
	replies that are not desired (hopefully helping to
	dramatically increase the diversity of search replies
	obtained over time).  Note that the actual encoding
	and databases are not affected (just P2P protocol).
	The update is not complete yet, but should compile.

Tue Jan  8 20:07:20 MST 2008
	Added option for testing ("make check") to use weak(er)
	PRNG for key generation (thanks to Werner Koch for 
	pointing out how to do it).

Thu Dec 22 20:10:37 MST 2007
	Releasing GNUnet 0.7.3.

Sun Dec  9 14:34:32 MST 2007
	Implemented MySQL version of the dstore-module.  This
	means that sqLite is now again truly optional.

Sat Dec  8 15:15:53 MST 2007
	Integrated F2F topology into main topology module 
	(options are used to choose between F2F-only (old F2F), 
	 minimum number of friend connections (NEW) or entirely
	 arbitrary connection set (default)).

Thu Dec  6 20:51:17 MST 2007
	The SMTP transport is working again.

Thu Nov 22 20:49:16 MST 2007
	United all libgnunetutil_*.so libraries into one big
	libgnunetutil.so library (eliminates issues for binary

src/applications/fs/tools/gnunet-download.c
/*
     This file is part of GNUnet.
     (C) 2001, 2002, 2003, 2004, 2006 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 2, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file applications/fs/tools/gnunet-download.c
 * @brief Main function to download files from GNUnet.
 * @author Christian Grothoff
 */

#include "platform.h"
#include "gnunet_directories.h"
#include "gnunet_fsui_lib.h"
#include "gnunet_util.h"

static struct GNUNET_GE_Context *ectx;

static struct GNUNET_GC_Configuration *cfg;

static unsigned long long verbose;

static int do_recursive;

static int do_directory;

static int do_delete_incomplete;

static char *cfgFilename = GNUNET_DEFAULT_CLIENT_CONFIG_FILE;

static char *filename;

static unsigned int anonymity = 1;

static unsigned int parallelism = 32;

static GNUNET_CronTime start_time;

static struct GNUNET_FSUI_DownloadList *dl;

#define EC_ARGUMENTS -1
#define EC_COMPLETED 0
#define EC_INCOMPLETE 1
#define EC_ABORTED 2
#define EC_DOWNLOAD_ERROR 3
#define DEBUG 0
static int errorCode;

static unsigned int downloads_running;

static struct GNUNET_FSUI_DownloadList **downloads;

static unsigned int downloads_size;

static struct GNUNET_Mutex *lock;

/**
 * All gnunet-download command line options
 */
static struct GNUNET_CommandLineOption gnunetdownloadOptions[] = {
  {'a', "anonymity", "LEVEL",
   gettext_noop ("set the desired LEVEL of sender-anonymity"),
   1, &GNUNET_getopt_configure_set_uint, &anonymity},
  GNUNET_COMMAND_LINE_OPTION_CFG_FILE (&cfgFilename),   /* -c */
  {'d', "directory", NULL,
   gettext_noop
   ("download a GNUnet directory that has already been downloaded.  Requires that a filename of an existing file is specified instead of the URI.  The download will only download the top-level files in the directory unless the `-R' option is also specified."),
   0, &GNUNET_getopt_configure_set_one, &do_directory},
  {'D', "delete-incomplete", NULL,
   gettext_noop ("delete incomplete downloads (when aborted with CTRL-C)"),
   0, &GNUNET_getopt_configure_set_one, &do_delete_incomplete},
  GNUNET_COMMAND_LINE_OPTION_HELP (gettext_noop ("Download files from GNUnet.")),       /* -h */
  GNUNET_COMMAND_LINE_OPTION_HOSTNAME,  /* -H */
  GNUNET_COMMAND_LINE_OPTION_LOGGING,   /* -L */
  {'o', "output", "FILENAME",
   gettext_noop ("write the file to FILENAME"),
   1, &GNUNET_getopt_configure_set_string, &filename},
  {'p', "parallelism", "DOWNLOADS",
   gettext_noop
   ("set the maximum number of parallel downloads that are allowed"),
   1, &GNUNET_getopt_configure_set_uint, &parallelism},
  {'R', "recursive", NULL,
   gettext_noop ("download a GNUnet directory recursively"),
   0, &GNUNET_getopt_configure_set_one, &do_recursive},
  GNUNET_COMMAND_LINE_OPTION_VERSION (PACKAGE_VERSION), /* -v */
  GNUNET_COMMAND_LINE_OPTION_VERBOSE,
  GNUNET_COMMAND_LINE_OPTION_END,
};

/**
 * This method is called whenever data is received.
 * The current incarnation just ensures that the main
 * method exits once the download is complete.
 */
static void *
progressModel (void *unused, const GNUNET_FSUI_Event * event)
{
  GNUNET_mutex_lock (lock);
  switch (event->type)
    {
    case GNUNET_FSUI_download_progress:
      if (verbose)
        {
          PRINTF (_("Download of file `%s' at "
                    "%16llu out of %16llu bytes (%8.3f KiB/s)\n"),
                  event->data.DownloadProgress.filename,
                  event->data.DownloadProgress.completed,
                  event->data.DownloadProgress.total,
                  (event->data.DownloadProgress.completed / 1024.0) /
                  (((double) (GNUNET_get_time () - (start_time - 1)))
                   / (double) GNUNET_CRON_SECONDS));
        }
      break;
    case GNUNET_FSUI_download_aborted:
      if (dl == event->data.DownloadError.dc.pos)
        {
          /* top-download aborted */
          printf (_("Download aborted.\n"));
          errorCode = EC_ABORTED;
          GNUNET_shutdown_initiate ();
        }
      break;
    case GNUNET_FSUI_download_error:
      printf (_("Error downloading: %s\n"),
              event->data.DownloadError.message);
      errorCode = EC_DOWNLOAD_ERROR;
      GNUNET_shutdown_initiate ();
      break;
    case GNUNET_FSUI_download_completed:
      PRINTF (_("Download of file `%s' complete.  "
                "Speed was %8.3f KiB per second.\n"),
              event->data.DownloadCompleted.filename,
              (event->data.DownloadCompleted.total / 1024.0) /
              (((double) (GNUNET_get_time () - (start_time - 1)))
               / (double) GNUNET_CRON_SECONDS));
      downloads_running--;
      if (downloads_running == 0)
        {
          errorCode = 0;
          GNUNET_shutdown_initiate ();
        }
      break;
    case GNUNET_FSUI_download_started:
      downloads_running++;
      GNUNET_array_append (downloads, downloads_size,
                           event->data.DownloadStarted.dc.pos);
    case GNUNET_FSUI_download_stopped:
      break;
    default:
      GNUNET_GE_BREAK (ectx, 0);
      break;
    }
  GNUNET_mutex_unlock (lock);
  return NULL;
}

static int
directoryIterator (const GNUNET_ECRS_FileInfo * fi,
                   const GNUNET_HashCode * key, int isRoot, void *cls)
{
  struct GNUNET_FSUI_Context *ctx = cls;
  struct GNUNET_MetaData *meta;
  char *fn;
  char *f;

  f = GNUNET_meta_data_get_first_by_types (fi->meta,
                                           EXTRACTOR_FILENAME,
                                           EXTRACTOR_TITLE,
                                           EXTRACTOR_ARTIST,
                                           EXTRACTOR_AUTHOR,
                                           EXTRACTOR_PUBLISHER,
                                           EXTRACTOR_CREATOR,
                                           EXTRACTOR_PRODUCER,
                                           EXTRACTOR_UNKNOWN, -1);
  if (f == NULL)
    f = GNUNET_strdup (_("no name given"));
  fn = GNUNET_malloc (strlen (filename) + strlen (f) + 4);
  strcpy (fn, filename);
  strcat (fn, "/");
  strcat (fn, f);
  if (verbose > 1)
    printf (_("Starting download `%s'\n"), f);
  GNUNET_free (f);
  meta = GNUNET_meta_data_create ();
  GNUNET_FSUI_download_start (ctx,
                              anonymity, do_recursive, fi->uri, meta, fn,
                              NULL, NULL);
  GNUNET_meta_data_destroy (meta);
  GNUNET_free (fn);
  return GNUNET_OK;
}


/**
 * Main function to download files from GNUnet.
 *
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return return value from download file: 0: ok, -1, 1: error
 */
int
main (int argc, char *const *argv)
{
  int ok;
  int try_rename;
  struct GNUNET_FSUI_Context *ctx;
  struct GNUNET_ECRS_URI *uri;
  struct GNUNET_MetaData *meta;
  int i;
#if DEBUG
  GNUNET_EncName *enc;
#endif

  i = GNUNET_init (argc,
                   argv,
                   "gnunet-download [OPTIONS] URI",
                   &cfgFilename, gnunetdownloadOptions, &ectx, &cfg);
  if (i == -1)
    {
      errorCode = EC_ARGUMENTS;
      goto quit;
    }
  if (i == argc)
    {
      GNUNET_GE_LOG (ectx,
                     GNUNET_GE_WARNING | GNUNET_GE_BULK | GNUNET_GE_USER,
                     _("Not enough arguments. "
                       "You must specify a GNUnet file URI\n"));
      errorCode = EC_ARGUMENTS;
      goto quit;
    }
  GNUNET_GC_get_configuration_value_number (cfg,
                                            "GNUNET",
                                            "VERBOSE", 0, 9999, 0, &verbose);
  uri = NULL;
  if (!do_directory)
    {
      uri = GNUNET_ECRS_string_to_uri (ectx, argv[i]);
#if DEBUG
      if (GNUNET_ECRS_uri_test_loc (uri))
        {
          enc = GNUNET_malloc (sizeof (GNUNET_EncName));
          GNUNET_ECRS_locURI_extract_peer (uri, &enc);
          fprintf (stdout,
                   "gnunet-download got locURI. Data located at peer %s.\n",
                   (char *) enc);
          GNUNET_free (enc);
        }
#endif
      if ((NULL == uri) ||
          (!(GNUNET_ECRS_uri_test_loc (uri)
             || GNUNET_ECRS_uri_test_chk (uri))))
        {
          GNUNET_GE_LOG (ectx,
                         GNUNET_GE_ERROR | GNUNET_GE_BULK | GNUNET_GE_USER,
                         _("URI `%s' invalid for gnunet-download.\n"),
                         argv[i]);
          errorCode = EC_ARGUMENTS;
          goto quit;
        }
    }

  try_rename = GNUNET_NO;
  if (filename == NULL)
    {
      if (do_directory)
        {
          if (NULL != strstr (argv[i], GNUNET_DIRECTORY_EXT))
            {
              filename = GNUNET_strdup (argv[i]);
              strstr (filename, GNUNET_DIRECTORY_EXT)[0] = '\0';
            }
          else
            {
              filename =
                GNUNET_malloc (strlen (argv[i]) +
                               strlen (GNUNET_DIRECTORY_EXT) + 2);
              strcpy (filename, argv[i]);
              strcat (filename, DIR_SEPARATOR_STR);
              strcat (filename, GNUNET_DIRECTORY_EXT);
            }
          try_rename = GNUNET_NO;
        }
      else
        {
          GNUNET_GE_ASSERT (ectx,
                            strlen (argv[i]) >
                            strlen (GNUNET_ECRS_URI_PREFIX) +
                            strlen (GNUNET_ECRS_FILE_INFIX));
          filename =
            GNUNET_expand_file_name (ectx,
                                     &argv[i][strlen (GNUNET_ECRS_URI_PREFIX)
                                              +
                                              strlen
                                              (GNUNET_ECRS_FILE_INFIX)]);
          GNUNET_GE_LOG (ectx,
                         GNUNET_GE_DEBUG | GNUNET_GE_REQUEST | GNUNET_GE_USER,
                         _
                         ("No filename specified, using `%s' instead (for now).\n"),
                         filename);
          try_rename = GNUNET_YES;
        }
    }
  ok = GNUNET_NO;
  lock = GNUNET_mutex_create (GNUNET_NO);
  ctx = GNUNET_FSUI_start (ectx,
                           cfg,
                           "gnunet-download",
                           parallelism == 0 ? 1 : parallelism,
                           GNUNET_NO, &progressModel, NULL);
  start_time = GNUNET_get_time ();
  errorCode = EC_INCOMPLETE;
  if (do_directory)
    {
      void *data;
      struct stat sbuf;
      int fd;
      int count;
      char *efn;

      fd = -1;
      efn = GNUNET_expand_file_name (ectx, argv[i]);
      data = NULL;
      if ((0 != STAT (efn,
                      &sbuf)) ||
          (!S_ISREG (sbuf.st_mode)) ||
          (0 != ACCESS (efn,
                        R_OK)) ||
          (-1 == (fd = GNUNET_disk_file_open (ectx,
                                              efn,
                                              O_LARGEFILE | O_RDONLY))) ||
          (MAP_FAILED == (data = MMAP (NULL,
                                       sbuf.st_size,
                                       PROT_READ, MAP_SHARED, fd, 0))))
        {
          if (fd != -1)
            CLOSE (fd);
          GNUNET_GE_LOG (ectx,
                         GNUNET_GE_ERROR | GNUNET_GE_IMMEDIATE |
                         GNUNET_GE_USER,
                         _("Could not access gnunet-directory file `%s'\n"),
                         efn);
          GNUNET_FSUI_stop (ctx);
          GNUNET_mutex_destroy (lock);
          GNUNET_free (efn);
          goto quit;
        }
      meta = GNUNET_meta_data_create ();
      count = GNUNET_ECRS_directory_list_contents (ectx,
                                                   data,
                                                   sbuf.st_size,
                                                   NULL,
                                                   &meta, &directoryIterator,
                                                   ctx);
      GNUNET_meta_data_destroy (meta);
      MUNMAP (data, sbuf.st_size);
      CLOSE (fd);
      GNUNET_free (efn);
      if (verbose > 0)
        {
          if (count > 0)
            printf (_("Downloading %d files from directory `%s'.\n"),
                    count, argv[i]);
          else
            printf (_("Did not find any files in directory `%s'\n"), argv[i]);
        }
    }
  else
    {
      meta = GNUNET_meta_data_create ();
      dl = GNUNET_FSUI_download_start (ctx,
                                       anonymity,
                                       do_recursive, uri, meta, filename,
                                       NULL, NULL);
      GNUNET_meta_data_destroy (meta);
      if (dl == NULL)
        {
          GNUNET_FSUI_stop (ctx);
          GNUNET_mutex_destroy (lock);
          goto quit;
        }
    }
  GNUNET_shutdown_wait_for ();
  if (do_delete_incomplete)
    {
      for (i = 0; i < downloads_size; i++)
        GNUNET_FSUI_download_abort (downloads[i]);
    }
  for (i = 0; i < downloads_size; i++)
    GNUNET_FSUI_download_stop (downloads[i]);
  GNUNET_array_grow (downloads, downloads_size, 0);
  GNUNET_FSUI_stop (ctx);
  GNUNET_mutex_destroy (lock);

  if ((errorCode == EC_COMPLETED) &&
      (dl != NULL) && (try_rename == GNUNET_YES))
    {
      char *newname = GNUNET_ECRS_suggest_better_filename (ectx,
                                                           filename);

      if (newname != NULL)
        {
          fprintf (stdout, _("File stored as `%s'.\n"), newname);
          GNUNET_free (newname);
        }
    }
  GNUNET_free (filename);
  if (uri != NULL)
    GNUNET_ECRS_uri_destroy (uri);
quit:
  GNUNET_fini (ectx, cfg);
  return errorCode;
}

/* end of gnunet-download.c */
